logviewer/api/transport_linux.go

```diff
@@ -0,0 +1,81 @@
+// build +linux
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"golang.org/x/sys/unix"
+	"net"
+	"os"
+	"runtime"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const titusInits = "/var/lib/titus-inits"
+
+// nsDialer is a net.Dialer that does a Setns() into a container's network namespace before doing a connect
+type nsDialer struct {
+	containerNetNsPath string
+	systemDialer       net.Dialer
+}
+
+func (d *nsDialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	log.Debugf("nsDialer: DialContext: %s", d.containerNetNsPath)
+	curThreadNsPath := fmt.Sprintf("/proc/%d/task/%d/ns/net", os.Getpid(), unix.Gettid())
+
+	// Make sure that this code executes only in this thread, since `Setns()` changes the thread's namespace
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// Get the FD for the host namespace, so we can switch back into it after doing the connect
+	prevNsFile, err := os.Open(curThreadNsPath)
```

Comment:
Golang x/unix has an Open API that might be slightly nicer to consume: https://godoc.org/golang.org/x/sys/unix#Open